### ISSUE TITLE
DDSP components in TorchAudio

### ISSUE THREAD

**Issue by mthrok** [pytorch/audio#2835]
Differential Digital Signal Processing (DDSP) is a technique proposed by the Google Magenta team. [[repo](https://github.com/magenta/ddsp), [papers](https://github.com/magenta/ddsp/tree/main/ddsp/training/gin/papers)].

It provides powerful generative methods that take advantage of the structure of sound based on Fourier analysis.
The paper has over 200 citations and is being applied to different tasks like speech synthesis and source separation.

TorchAudio is looking to support DDSP-based modeling methods by adding basic components.
We would like to hear and get help from the community on this work stream.

## Initial Goal
To support DDSP-based generative modeling. 

## New Components / APIs
1. Harmonic Synthesizer / Additive Synthesis
    - Differentiable harmonic synthesizer
        - WIP: oscillator bank [[prototype](https://github.com/mthrok/audio/blob/83ab22bc7275e51a28bcef9930a1b29e8252589c/torchaudio/prototype/functional/_dsp.py#L8)]
        - WIP: Harmonic overtones [[prototype](https://github.com/mthrok/audio/blob/83ab22bc7275e51a28bcef9930a1b29e8252589c/torchaudio/prototype/functional/_dsp.py#L87)]
    - WIP: [Optional] ADSR envelope generator [[prototype](https://github.com/mthrok/audio/blob/83ab22bc7275e51a28bcef9930a1b29e8252589c/torchaudio/prototype/functional/_dsp.py#L50)]
1. Subtractive Synthesis
    - Filter design functions
        - WIP: windowed-sinc filter [prototype] [[reference impl](https://github.com/magenta/ddsp/blob/7cb3c37f96a3e5b4a2b7e94fdcc801bfd556021b/ddsp/core.py#L1576)]
        - frequency method [[reference impl](https://github.com/magenta/ddsp/blob/7cb3c37f96a3e5b4a2b7e94fdcc801bfd556021b/ddsp/core.py#L1534)]
    - WIP: time-varying filter application function [[prototype](https://github.com/mthrok/audio/blob/83ab22bc7275e51a28bcef9930a1b29e8252589c/torchaudio/prototype/functional/_dsp.py#L159)] [[reference impl](https://github.com/magenta/ddsp/blob/7cb3c37f96a3e5b4a2b7e94fdcc801bfd556021b/ddsp/core.py#L1382-L1473)]
1. Reverb
   - Function to apply IR to waveform (stationary version of filter application function)

## Components for DDSP Modeling from [1]
1. Multi-scale mel spectrogram loss
  Used to train models. Compare the melspectrogram of the original waveform and synthesized waveform at multiple scales. [[reference impl](https://github.com/magenta/ddsp/blob/9866d3aaa6d13e3b951d39d5979d3b29697067ea/ddsp/losses.py#L132)]
1. exp_sigmoid
Modified sigmoid function used to ensure that model output is in the value range expected by synthesizers [[reference impl](https://github.com/magenta/ddsp/blob/7cb3c37f96a3e5b4a2b7e94fdcc801bfd556021b/ddsp/core.py#L386-L404)]
1. [Nice to have] Dataset
    - [NSynth](https://magenta.tensorflow.org/datasets/nsynth)
    - Solo Violin
1. [Stretch] DDSP Model from [1]
    - AutoEncoder model for Timbre transfer
1. [Stretch] Training Script and pre-trained model
    Script to train models for
    - LJSpeech
    - Solo violin

## Existing Components / Workstream
- [convolve](https://pytorch.org/audio/main/prototype.functional.html#convolve) and [fftconvolve](https://pytorch.org/audio/main/prototype.functional.html#fftconvolve) will be the basis for filter application.
- [RIR simulation](https://github.com/pytorch/audio/pull/2644) can be used for reverb.

## References
1. [[2001.04643] DDSP: Differentiable Digital Signal Processing](https://arxiv.org/abs/2001.04643)

 and some random papers moto found interesting
1. [[2010.15084] Speech Synthesis and Control Using Differentiable DSP](https://arxiv.org/abs/2010.15084)
1. [[2202.00200] Differentiable Digital Signal Processing Mixture Model for Synthesis Parameter Extraction from Mixture of Harmonic Sounds](https://arxiv.org/abs/2202.00200)
1. [[2210.14476] Sinusoidal Frequency Estimation by Gradient Descent](https://arxiv.org/abs/2210.14476)
